ports/raspberrypi/common-hal/pulseio/PulseIn.c

```diff
@@ -0,0 +1,231 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2017-2018 Scott Shawcroft for Adafruit Industries
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "src/rp2_common/hardware_gpio/include/hardware/gpio.h"
+#include "src/rp2_common/hardware_irq/include/hardware/irq.h"
+
+#include <stdint.h>
+
+#include "background.h"
+#include "mpconfigport.h"
+#include "py/gc.h"
+#include "py/runtime.h"
+#include "shared-bindings/microcontroller/__init__.h"
+#include "shared-bindings/pulseio/PulseIn.h"
+#include "shared-bindings/microcontroller/Pin.h"
+#include "supervisor/shared/translate.h"
+#include "bindings/rp2pio/StateMachine.h"
+#include "common-hal/pulseio/PulseIn.h"
+
+pulseio_pulsein_obj_t* save_self;
+
+#define NO_PIN 0xff
+
+const uint16_t pulsein_program[] = {
+//   set    pindirs, 0 ; For input
+    0xe080,
+//   wait   0 pin, 0   ; Wait for first low to start
+    0x2020,
+//   irq wait 0        ; set IRQ 0 and wait
+    0xc020,
+//   .bitloop
+//   in    pins, 1 [1] ; sample every 3 cycles (2 instructions, 1 delay)
+    0x4101,
+//   jmp bitloop
+    0x0003,
```

Comment:
I should explain the rationale behind the frequency and delay choices. Since pulsein wants to know the length of pulses (in microseconds), and since I couldn't get PIO to count properly, I am sampling every 4 us or so, and pushing those bits into a 32-bit word. Not trying to get every bit, just enough of a sampling so that in the interrupt handler a calculation can be done to give the desired result. In this case, I am multiplying the counted bits for a pulse by 6. The result is typically within a couple microseconds of what the logic analyzer is showing for the actual time. The frequency and delay times were figured out mostly by trail and error; if there are better ways to accomplish the task I would be glad to attempt implementing them.

Reply:
> The frequency and delay times were figured out mostly by trial and error; if there are better ways to accomplish the task I would be glad to attempt implementing them.

I think your definitely on the right track. Most implementations only get within a few us at best anyway. I'd choose a frequency that is an even number of microseconds to make the math simpler.

Reply:
Working on moving to an interrupt per word basis. I suspect this will affect the necessary frequency, as the SM program will change.

Reply:
Current version no longer needs delays - interrupts on a per-word basis seem to have good timing without them.